### ISSUE TITLE
Building for armv7 and arm64 instructions

### ISSUE THREAD

**Issue by guidedways** [libical/libical#187]
Can we please have the wiki include instructions to build for armv7 and arm64 devices?

**Comment by ahalls** [comment]
I've created a project, libical/XbICalendar, that builds for armv7 and arm64 devices using Apple's xcode.  We could build instructions out of that, or conversily if there is a better way to build for these platforms incorporate this process in XBlCalendar

**Comment by guidedways** [comment]
Thanks for sharing that - the static library you built, what version of libical did you use?

**Comment by guidedways** [comment]
Ah I see you've used libical 1.00 from sourceforge. What I want to be able to do is build the latest bleeding edge version of libical which now uses cmake. I've tried using ios-toolchain for cmake but they're all out of date and broken.

**Comment by guidedways** [comment]
Isn't ideal but I was able to make it the actual source code work. First I built it on a x64 platform, took all the source files from under "src", added it straight to our iOS project. Made minor adjustments to config.h by disabling ICU and voila. Worked.

**Comment by winterz** [comment]
I created a wiki page to put build instructions at https://github.com/libical/libical/wiki/Build-Instructions
and welcome definitive instructions for building on arm to be added there.

**Comment by ahalls** [comment]
We have built the library using cmake and the tag 1.0.1 for libical.  The build scripts are in the repro so you can modify them to build the bleeding edge if you would like.

**Comment by yakaolife** [comment]
Hi, I've managed to build libical for iOS development using CMake, thought I'd shared what I did and maybe people/author can also comment on this approach.

For armv7, I added the following inside CMakeList.txt:
SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -arch armv7")
SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -arch armv7”)

Then run cmake ..

Modify the following line from the build/CMakeCache.txt that's generated by cmake command:

//The product will be built against the headers and libraries located
// inside the indicated SDK.
CMAKE_OSX_SYSROOT:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk

(Change it into iPhoneOS.sdk, or iPhoneOS8.4.sdk for me)

run make, and make install as noted in the installation from libical.

use lipo -info to check the generated lib .a files, it will say it is armv7. 

I use the same way to generate .a for arm64 and also i386 (for different simulators, etc), and use lipo -create to combine all of the into one FAT.
(Let me know if this is the wrong place to post!)

**Comment by winterz** [comment]
Thanks for the feedback Betsy.

I will add these notes to our wiki pages at https://github.com/libical/libical/wiki/For-the-Users:-Build-for-Different-Platforms

**Comment by winterz** [comment]
I just found http://stackoverflow.com/questions/12630970/compiling-for-ios-with-cmake and the ios-cmake toolchain file.

could you check to see if the instructions using the ios-cmake toolchain file will work for you as shown in the answer at http://stackoverflow.com/questions/12630970/compiling-for-ios-with-cmake

looks like you'll need CMake version 3.0 or higher

**Comment by winterz** [comment]
1 year old and I'm assuming the OP is satisfied.  closing this.

**Comment by henrybear327** [comment]
Hi all, 

I tried adding `SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -arch armv7")` and `SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -arch armv7”)`, and also change the path to `CMAKE_OSX_SYSROOT:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk`. But I am getting a lot of `error: Unsupported architecture` during `make`

Some errors are like the one shown below.
```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
#error Unsupported architecture
```

May I know if anyone has recently successfully built the lib for iOS? 

Thank you!

**Comment by henrybear327** [comment]
Hi all,
I tried adding `SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -arch armv7")` and `SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -arch armv7”)`, and also change the path to `CMAKE_OSX_SYSROOT:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk`. But I am getting a lot of `error: Unsupported architecture during make`

Some errors are like the one shown below.
```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
#error Unsupported architecture
```

So I tried to change the `Toolchain-iOS.cmake` file to [this](https://github.com/cristeab/ios-cmake/blob/master/toolchain/iOS.cmake) one and compiled with `cmake -DENABLE_GTK_DOC=OFF -DCMAKE_TOOLCHAIN_FILE=/Users/xxx/Downloads/libical/cmake/Toolchain-iOS.cmake -DPERL_EXECUTABLE=/usr/bin/perl -DPKG_CONFIG_EXECUTABLE=/usr/local/bin/pkg-config ..`, and then some linker error happened. 
```
Scanning dependencies of target src-generator
[ 55%] Building C object src/libical-glib/CMakeFiles/src-generator.dir/tools/generator.c.o
[ 55%] Building C object src/libical-glib/CMakeFiles/src-generator.dir/tools/xml-parser.c.o
[ 56%] Linking C executable ../../bin/src-generator
ld: warning: -headerpad_max_install_names is ignored when used with -bitcode_bundle (Xcode setting ENABLE_BITCODE=YES)
ld: warning: ignoring file /usr/local/Cellar/glib/2.64.4/lib/libgobject-2.0.dylib, building for iOS-armv7 but attempting to link with file built for macOS-x86_64
ld: warning: ignoring file /usr/local/Cellar/glib/2.64.4/lib/libglib-2.0.dylib, building for iOS-armv7 but attempting to link with file built for macOS-x86_64
ld: warning: ignoring file /usr/local/opt/gettext/lib/libintl.dylib, building for iOS-armv7 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture armv7:
  "_g_ascii_strncasecmp", referenced from:
      _generate_library in generator.c.o
  "_g_build_filename", referenced from:
      _open_file in generator.c.o
      _generate_library in generator.c.o
  "_g_clear_error", referenced from:
      _generate_library in generator.c.o
  "_g_dir_close", referenced from:
      _generate_library in generator.c.o
  "_g_dir_open", referenced from:
      _generate_library in generator.c.o
  "_g_dir_read_name", referenced from:
      _generate_library in generator.c.o
  "_g_file_test", referenced from:
      _generate_library in generator.c.o
  "_g_free", referenced from:
      _get_source_method_comment in generator.c.o
      _get_upper_camel_from_lower_snake in generator.c.o
      _get_upper_snake_from_lower_snake in generator.c.o
      _get_upper_snake_from_upper_camel in generator.c.o
      _get_lower_snake_from_upper_camel in generator.c.o
      _generate_header_method_get_type in generator.c.o
      _generate_header_method_proto in generator.c.o
      ...
  "_g_hash_table_contains", referenced from:
      _generate_code_from_template in generator.c.o
      _generate_conditional in generator.c.o
      _get_source_method_body in generator.c.o
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_translator_for_parameter in generator.c.o
      ...
  "_g_hash_table_destroy", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _generate_header_and_source in generator.c.o
      _generate_library in generator.c.o
      _structure_free in xml-parser.c.o
  "_g_hash_table_insert", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_header_enum in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      ...
  "_g_hash_table_iter_init", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
  "_g_hash_table_iter_next", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
  "_g_hash_table_lookup", referenced from:
      _generate_code_from_template in generator.c.o
      _generate_conditional in generator.c.o
      _get_source_method_body in generator.c.o
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_translator_for_parameter in generator.c.o
      ...
  "_g_hash_table_new_full", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      _structure_new in xml-parser.c.o
      ...
  "_g_list_append", referenced from:
      _generate_header_method_new_full in generator.c.o
      _get_source_method_proto_new_full in generator.c.o
      _get_source_method_proto_set_owner in generator.c.o
      _get_source_method_proto_set_native in generator.c.o
      _get_source_method_proto_set_is_global in generator.c.o
      _get_source_method_proto_set_property in generator.c.o
      _get_source_method_proto_get_property in generator.c.o
      ...
  "_g_list_first", referenced from:
      _get_source_method_comment in generator.c.o
      _generate_header_method_proto in generator.c.o
      _get_source_method_proto in generator.c.o
      _generate_header_method_protos in generator.c.o
      _generate_code_from_template in generator.c.o
      _get_source_method_body in generator.c.o
      _generate_header_enums in generator.c.o
      ...
  "_g_list_free", referenced from:
      _structure_free in xml-parser.c.o
      _method_free in xml-parser.c.o
      _enumeration_free in xml-parser.c.o
      _parameter_free in xml-parser.c.o
      _ret_free in xml-parser.c.o
  "_g_list_free_full", referenced from:
      _generate_forward_declarations_header_file in generator.c.o
      _generate_library in generator.c.o
  "_g_list_last", referenced from:
      _generate_code_from_template in generator.c.o
  "_g_list_prepend", referenced from:
      _generate_forward_declarations_header_file in generator.c.o
      _generate_library in generator.c.o
  "_g_list_sort", referenced from:
      _generate_forward_declarations_header_file in generator.c.o
      _generate_library in generator.c.o
  "_g_log", referenced from:
      _write_str in generator.c.o
      _get_inline_parameter in generator.c.o
      _generate_header_enum in generator.c.o
      _generate_library in generator.c.o
  "_g_malloc", referenced from:
      _get_source_method_proto in generator.c.o
      _get_source_method_code in generator.c.o
  "_g_malloc0_n", referenced from:
      _structure_new in xml-parser.c.o
      _declaration_new in xml-parser.c.o
      _method_new in xml-parser.c.o
      _parameter_new in xml-parser.c.o
      _ret_new in xml-parser.c.o
      _enumeration_new in xml-parser.c.o
  "_g_malloc_n", referenced from:
      _get_source_method_comment in generator.c.o
      _get_upper_camel_from_lower_snake in generator.c.o
      _get_upper_snake_from_lower_snake in generator.c.o
      _get_lower_snake_from_upper_camel in generator.c.o
      _generate_header_method_proto in generator.c.o
      _get_source_method_proto in generator.c.o
      _generate_code_from_template in generator.c.o
      ...
  "_g_printerr", referenced from:
      _get_translator_for_parameter in generator.c.o
      _get_translator_for_return in generator.c.o
  "_g_return_if_fail_warning", referenced from:
      _get_source_method_comment in generator.c.o
      _get_upper_camel_from_lower_snake in generator.c.o
      _get_upper_snake_from_lower_snake in generator.c.o
      _get_upper_snake_from_upper_camel in generator.c.o
      _get_lower_snake_from_upper_camel in generator.c.o
      _get_lower_train_from_lower_snake in generator.c.o
      _get_lower_train_from_upper_camel in generator.c.o
      ...
  "_g_stpcpy", referenced from:
      _get_source_method_comment in generator.c.o
      _get_source_method_proto in generator.c.o
      _generate_conditional in generator.c.o
      _get_source_method_body in generator.c.o
      _get_source_method_code in generator.c.o
      _get_inline_parameter in generator.c.o
      _get_source_run_time_checkers in generator.c.o
      ...
  "_g_str_equal", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      _structure_new in xml-parser.c.o
      ...
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      _structure_new in xml-parser.c.o
      ...
  "_g_str_hash", referenced from:
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      _structure_new in xml-parser.c.o
      ...
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      _generate_forward_declarations_header_file in generator.c.o
      _get_hash_table_from_structure in generator.c.o
      _generate_library in generator.c.o
      _initialize_default_value_table in generator.c.o
      _structure_new in xml-parser.c.o
      ...
  "_g_strchomp", referenced from:
      _get_list_from_string in xml-parser.c.o
  "_g_strchug", referenced from:
      _get_list_from_string in xml-parser.c.o
  "_g_strcmp0", referenced from:
      _generate_header_method_protos in generator.c.o
      _open_private_header in generator.c.o
      _generate_code_from_template in generator.c.o
      _generate_conditional in generator.c.o
      _get_source_method_body in generator.c.o
      _generate_header_includes in generator.c.o
      _generate_source_includes in generator.c.o
      ...
  "_g_strconcat", referenced from:
      _get_source_method_comment in generator.c.o
      _generate_header_method_get_type in generator.c.o
      _generate_header_method_new_full in generator.c.o
      _get_source_method_proto_new_full in generator.c.o
      _get_source_method_proto_set_owner in generator.c.o
      _get_source_method_proto_set_native in generator.c.o
      _get_source_method_proto_set_is_global in generator.c.o
      ...
  "_g_strdup", referenced from:
      _get_source_method_comment in generator.c.o
      _get_lower_snake_from_upper_camel in generator.c.o
      _get_lower_train_from_lower_snake in generator.c.o
      _generate_header_method_get_type in generator.c.o
      _generate_header_method_new_full in generator.c.o
      _get_source_method_proto_new_full in generator.c.o
      _get_source_method_proto_set_owner in generator.c.o
      ...
  "_g_strerror", referenced from:
      _write_str in generator.c.o
  "_g_strsplit", referenced from:
      _get_list_from_string in xml-parser.c.o
  "_g_strv_length", referenced from:
      _get_list_from_string in xml-parser.c.o
ld: symbol(s) not found for architecture armv7
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/src-generator] Error 1
make[1]: *** [src/libical-glib/CMakeFiles/src-generator.dir/all] Error 2
make: *** [all] Error 2
```

May I know if anyone has recently successfully built the lib for iOS? I think I am missing some steps here for sure.

Thank you!

**Comment by kosuel** [comment]
I tried to build for iOS. But there is a error like below:
It seems that system() function is not available anymore with iOS SDK.

/Users/ohhyung/Downloads/libical-3.0/src/libicalss/icalfileset.c:370:13: error: 'system' is unavailable: not available on iOS
        if (system(tmp) < 0) {
            ^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/stdlib.h:184:6: note: 'system' has been explicitly marked unavailable here
int      system(const char *) __DARWIN_ALIAS_C(system);
         ^
1 error generated.
make[2]: *** [src/libicalss/CMakeFiles/icalss.dir/icalfileset.c.o] Error 1
make[1]: *** [src/libicalss/CMakeFiles/icalss.dir/all] Error 2
make: *** [all] Error 2

**Comment by winterz** [comment]
> I tried to build for iOS. But there is a error like below: It seems that system() function is not available anymore with iOS SDK.
> 
> ```
> _### /Users/ohhyung/Downloads/libical-3.0/src/libicalss/icalfileset.c:370:13: error: 'system' is unavailable: not available on iOS
>         if (system(tmp) < 0) {
>             ^
> /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/stdlib.h:184:6: note: 'system' has been explicitly marked unavailable here
> int      system(const char *) __DARWIN_ALIAS_C(system);
>          ^
> 1 error generated.
> make[2]: *** [src/libicalss/CMakeFiles/icalss.dir/icalfileset.c.o] Error 1
> make[1]: *** [src/libicalss/CMakeFiles/icalss.dir/all] Error 2
> make: *** [all] Error 2_
> ```

We were warned about this issue #312 and never did anything about it.  For now you could comment the offending code -- I doubt you need it.  we should put a real fix into the 3.1 release.

**Comment by groue** [comment]
Hello, I also met the `error: 'system' is unavailable: not available on iOS` error while trying to cross-compile libical for iOS.
